### ISSUE TITLE
MultiJoin transform warns of no "stable unique name"

### ISSUE THREAD

**Issue by echeran** [spotify/scio#466]
After upgrading Scio from 0.2.6 to 0.2.11 and keeping the application code the same, I am getting warnings coming from MultiJoin.outer complaining that a transform named CoGroupByKey2 has no stable unique name:

```
2017-02-22 18:08:57,708 WARN  com.spotify.scio.VersionUtil$            - A newer version of Scio is available: 0.2.11 -> v0.3.0-alpha2
2017-02-22 18:09:02,363 WARN  com.google.cloud.dataflow.sdk.Pipeline   - Transform transform@{LoaderPipeline.scala:249}/transform@{LoaderPipeline.scala:245}/CoGroupByKey2 does not have a stable unique name. This will prevent updating of pipelines.
2017-02-22 18:09:02,876 WARN  com.google.cloud.dataflow.sdk.Pipeline   - Transform transform@{LoaderPipeline.scala:190}/transform@{LoaderPipeline.scala:186}/CoGroupByKey2 does not have a stable unique name. This will prevent updating of pipelines.
2017-02-22 18:09:03,175 WARN  com.google.cloud.dataflow.sdk.Pipeline   - Transform CoGroupByKey2 does not have a stable unique name. This will prevent updating of pipelines.
```

where the error also surfaces in an integration test that gives a stacktrace that looks like this:

```
[info]   java.lang.IllegalStateException: Transform CoGroupByKey2 does not have a stable unique name. This will prevent updating of pipelines.
[info]   at com.google.cloud.dataflow.sdk.Pipeline.applyInternal(Pipeline.java:349)
[info]   at com.google.cloud.dataflow.sdk.Pipeline.applyTransform(Pipeline.java:291)
[info]   at com.google.cloud.dataflow.sdk.transforms.join.KeyedPCollectionTuple.apply(KeyedPCollectionTuple.java:116)
[info]   at com.spotify.scio.util.MultiJoin$class.outer(MultiJoin.scala:2304)
[info]   at com.spotify.scio.util.NamedMultiJoin.outer(MultiJoin.scala:2717)
[info]   at com.mycompany.mypackage.loader.agg.AggregationUtil$.aggregateSomeEventType(AggregationUtil.scala:432)
[info]   at com.mycompany.mypackage.loader.AggSomeEventTypePipeline$.com$nest$elixir$carat$loader$AggSomeEventTypePipeline$$testTransformWithLogInfos$1(DFEventAggTestPipelines.scala:43)
[info]   at com.mycompany.mypackage.loader.AggSomeEventTypePipeline$$anonfun$main$1.apply(DFEventAggTestPipelines.scala:46)
[info]   at com.mycompany.mypackage.loader.AggSomeEventTypePipeline$$anonfun$main$1.apply(DFEventAggTestPipelines.scala:46)
[info]   at com.mycompany.mypackage.loader.DFEventAggTestUtil$.mainWithLogInfos(DFEventAggTestUtil.scala:45)
[info]   ...

```

And the stacktrace goes through code that looks like this, where line 432 is the `MultiJoin.outer(` line:

```
    // join the SCollections back together
    val multiJoined =
      MultiJoin.outer(
        durationSColls(0),
        durationSColls(1),
        durationSColls(2),
        durationSColls(3),
        durationSColls(4),
        durationSColls(5),
        durationSColls(6),
        durationSColls(7),
        durationSColls(8),
        durationSColls(9),
        durationSColls(10),
        durationSColls(11),
        durationSColls(12),
        durationSColls(13)
      )

    val aggEvents = multiJoined map { tuple => {
      val (field1, field2, ...
           ) = tuple
      val aggEvent =
        AggEventType(
          field1, field2, ...
          )
      aggEvent
    }}
```

I think it's related to issue #354 and the related PR #357, but I wasn't able to trace through the call chain from application code to Scio code to further diagnose more precisely.  So instead, I tried removing `.setName()` calls on SCollections in the code, including on SColls inside `durationSColls` that is the input to the MultiJoin.outer call.  I also tried changing some of the code preceding of the the definition of `durationSColls` from `scoll1.map(someFn).setName("transform1")` to `scoll1.withName("transform1").map(someFn)`.  None of these changes in my application code avoided the warnings, leading me to speculate that it's something inside Scio.

**Comment by andrewsmartin** [comment]
Hey @echeran, thanks for reporting. I'm unable to reproduce using just this code snippet, but could you try setting a name for the multi-join yourself?

```
val multiJoined = MultiJoin.withName("MyOuterJoin").outer(...)
```

Scio internally assigns the name `CoGroupByKey` when performing multi-joins, and it looks like there is probably another `CoGroupByKey` somewhere else because the join is named `CoGroupByKey2` here. The number is added by Dataflow when there are colliding transform names. In any case, try giving it a name yourself and see if that works.

Out of curiosity, is this just a warning, or is it preventing your job from running?

**Comment by andrewsmartin** [comment]
Also, just an FYI: `scoll1.map(someFn).setName("transform1")` actually sets the name of the `SCollection` itself, not the producing transform.

**Comment by andrewsmartin** [comment]
I just confirmed that I get the same warning by having another `CoGroupByKey` before the multi-join. This warning is expected behavior, because Dataflow is warning that changes to your code may cause the transform name to change which would make it harder to update (if it was a streaming job). However, I tried setting the names manually using `MultiJoin.withName()` and still get the warning - this is definitely a bug.
